src/cpp/rtps/history/TopicPayloadPool.hpp

```diff
@@ -0,0 +1,348 @@
+            const PoolConfig& config,
+            bool is_reserve);
+
+    /**
+     * Ensures the pool has capacity for at least @c num_payloads elements.
+     *
+     * @param [IN] min_num_payloads Minimum number of payloads reserved in the pool
+     * @param [IN] size             Size to allocate for the payloads that need to be added to the pool
+     *
+     * @pre
+     *   - @c min_num_payloads <= @c max_pool_size_
+     * @post
+     *   - @c payload_pool_allocated_size() >= @c min_num_payloads
+     */
+    virtual void reserve (
+            uint32_t min_num_payloads,
+            uint32_t size);
+
+    /**
+     * Ensures the pool has capacity for at most @c num_payloads elements.
+     *
+     * @param [IN] max_num_payloads Maximum number of payloads reserved in the pool
+     *
+     * @return @c true on success, @c false otherwise
+     *
+     * @post
+     *   - On success, payload_pool_allocated_size() <= max_num_payloads
+     *   - On failure, memory for some payloads may have been released, but payload_pool_allocated_size() > min_num_payloads
+     */
+    bool shrink (
+            uint32_t max_num_payloads);
+
+    /**
+     * @brief Get a serialized payload for a new sample.
+     *
+     * If the payload is recycled from the pool, @c resizable controls whether it can
+     * be reallocated to accomodate larger sizes.
+     * If @c resizable is false and there is at least one free payload in the pool, that payload will
+     * be returned even though it may not reach the requested size.
+     *
+     * If @c resizable is true and the reallocation fails, the operation returns false and
+     * the payload is returned to the pool.
+     *
+     * @param [in]     size          Number of bytes required for the serialized payload
+     * @param [in,out] cache_change  Cache change to assign the payload to
+     * @param [in]     resizable     Whether payloads recycled from the pool are resizable to accomodate larger sizes
+     *
+     * @returns whether the operation succeeded or not
+     *
+     * @post
+     *   On success:
+     *     @li Field @c cache_change.payload_owner equals this
+     *     @li Field @c serializedPayload.data points to a buffer of at least @c size bytes
+     *     @li Field @c serializedPayload.max_size is greater than or equal to @c size
+     */
+    virtual bool do_get_payload(
+            uint32_t size,
+            CacheChange_t& cache_change,
+            bool resizeable);
+
+    virtual MemoryManagementPolicy_t memory_policy() const = 0;
+
+    uint32_t max_pool_size_             = 0;  //< Maximum size of the pool
+    uint32_t infinite_histories_count_  = 0;  //< Number of infinite histories reserved
+    uint32_t finite_max_pool_size_      = 0;  //< Maximum size of the pool if no infinite histories were reserved
+
+    std::vector<PayloadNode*> free_payloads_; //< Payloads that are free
+    std::vector<PayloadNode*> all_payloads_;  //< All payloads
```

Comment:
If in TopicPayloadPool we are using `vector` instead of `ResourceLimitedVector` to avoid breaking the index position kept in the PayloadNode, we should provide a comment on it (specifying a *sequence container* is required). In fact, we are using **all_payload_** vector using the `ResourceLimitedVector strategy`.